### PR TITLE
Update password policy in sign up flow

### PR DESCRIPTION
**Pull request by ch5zzy**
Catch the PASSWORD_DOES_NOT_MEET_REQUIREMENTS error during sign up and update the password policy if it is already cached for the project/tenant.

### CODE REVIEW

File: packages/auth/src/core/strategies/email_and_password.test.ts [hunk 1]
```diff
@@ -745,6 +745,149 @@ describe('core/strategies/email_and_password/createUserWithEmailAndPassword', ()
...
+            error: {
+              code: 400,
+              message: ServerError.PASSWORD_DOES_NOT_MEET_REQUIREMENTS
+            }
+          },
+          400
+        );
+      });
+
+      it('updates the password policy when password does not meet backend requirements', async () => {
```
Comment (Xiaoshouzi-gh) [inline packages/auth/src/core/strategies/email_and_password.test.ts:842]: may be reword it to - updates the *cached* password policy when password does not meet backend requirements

File: packages/auth/src/core/strategies/email_and_password.test.ts [hunk 2]
```diff
@@ -745,6 +745,149 @@ describe('core/strategies/email_and_password/createUserWithEmailAndPassword', ()
...
+        await auth._updatePasswordPolicy();
+        expect(policyEndpointMock.calls.length).to.eq(1);
+        expect(auth._getPasswordPolicy()).to.eql(cachedPasswordPolicy);
+
```
Comment (Xiaoshouzi-gh) [inline packages/auth/src/core/strategies/email_and_password.test.ts:846]: Please add a comment here: policy changed after the previous fetch.

File: packages/auth/src/core/strategies/email_and_password.test.ts [hunk 3]
```diff
@@ -745,6 +745,149 @@ describe('core/strategies/email_and_password/createUserWithEmailAndPassword', ()
...
+          FirebaseError,
+          'Firebase: The password does not meet the requirements. (auth/password-does-not-meet-requirements).'
+        );
+
+        expect(policyEndpointMock.calls.length).to.eq(0);
+        expect(auth._getPasswordPolicy()).to.be.null;
+      });
+
+      it('does not update the password policy upon error if tenant changes and policy has not previously been fetched', async () => {
+        auth.tenantId = 'tenant-id';
```
Comment (Xiaoshouzi-gh) [inline packages/auth/src/core/strategies/email_and_password.test.ts:874]: string `tenant-id` shows up more than once, please consider putting it into a constant.

File: packages/auth/src/core/strategies/email_and_password.ts [hunk 4]
```diff
@@ -308,6 +308,15 @@ export async function createUserWithEmailAndPassword(
         );
         return signUp(authInternal, requestWithRecaptcha);
       } else {
+        // Fetch the password policy if the password did not meet policy requirements and there is an existing policy cached.
```
Comment (Xiaoshouzi-gh) [inline packages/auth/src/core/strategies/email_and_password.ts:311]: Can you also add a comment here stating how would a customer be able to use this feature for the first time?

**Review by Xiaoshouzi-gh — APPROVED** [review]
LGTM

File: packages/auth/src/core/strategies/email_and_password.ts (hunk 4 above)
Comment (ch5zzy) [reply packages/auth/src/core/strategies/email_and_password.ts]: Done.

File: packages/auth/src/core/strategies/email_and_password.test.ts (hunk 3 above)
Comment (ch5zzy) [reply packages/auth/src/core/strategies/email_and_password.test.ts]: Done.

File: packages/auth/src/core/strategies/email_and_password.test.ts (hunk 2 above)
Comment (ch5zzy) [reply packages/auth/src/core/strategies/email_and_password.test.ts:852]: Done.

File: packages/auth/src/core/strategies/email_and_password.test.ts (hunk 1 above)
Comment (ch5zzy) [reply packages/auth/src/core/strategies/email_and_password.test.ts]: I see how this could be confusing. I reworded all of them to be consistent.

File: packages/auth/src/core/strategies/email_and_password.test.ts [hunk 5]
```diff
@@ -745,6 +745,149 @@ describe('core/strategies/email_and_password/createUserWithEmailAndPassword', ()
...
+          'Firebase: The password does not meet the requirements. (auth/password-does-not-meet-requirements).'
+        );
+
+        expect(policyEndpointMock.calls.length).to.eq(2);
+        expect(auth._getPasswordPolicy()).to.eql(
+          cachedPasswordPolicyRequireNumeric
+        );
+      });
+
+      it('does not update the password policy upon error if policy has not previously been fetched', async () => {
```
[the remark this answers is not in the document]
Comment (ch5zzy) [reply packages/auth/src/core/strategies/email_and_password.test.ts]: I added the comment in `email_and_password.ts` to have it by the logic this is testing. Done.

File: packages/auth/src/core/strategies/email_and_password.test.ts [hunk 6]
```diff
@@ -745,6 +745,149 @@ describe('core/strategies/email_and_password/createUserWithEmailAndPassword', ()
...
+        await expect(
+          createUserWithEmailAndPassword(auth, 'some-email', 'some-password')
+        ).to.be.rejectedWith(
+          FirebaseError,
+          'Firebase: The password does not meet the requirements. (auth/password-does-not-meet-requirements).'
```
[the remark this answers is not in the document]
Comment (ch5zzy) [reply packages/auth/src/core/strategies/email_and_password.test.ts]: Done.

**Comment by ch5zzy** [comment]
I thought of an error case where a `PASSWORD_DOES_NOT_MEET_REQUIREMENTS` error is received and the new policy fetched from the backend has an unsupported schema version. I discussed this with Chuan and we decided that to propagate the `UNSUPPORTED_PASSWORD_POLICY_SCHEMA_VERSION` error to the developer, we can store the schema version internally and check this value when `validatePassword` is called. I'll do this in a future PR but wanted to document this here since it is related to the auth flow.
